pkg/exchange/coinbase/stream.go

```diff
@@ -25,10 +37,156 @@ type Stream struct {
 	activeMessageCallbacks              []func(m *ActiveMessage)
 }
 
-func NewStream() *Stream {
+func NewStream(
+	apiKey string,
+	passphrase string,
+	secretKey string,
+) *Stream {
 	s := Stream{
 		StandardStream: types.NewStandardStream(),
+		apiKey:         apiKey,
+		passphrase:     passphrase,
+		secretKey:      secretKey,
 	}
 	s.SetParser(s.parseMessage)
+	s.SetDispatcher(s.dispatchEvent)
+	s.SetEndpointCreator(createEndpoint)
+
+	// public handlers
+	s.OnConnect(s.handleConnect)
+
 	return &s
 }
+
+func (s *Stream) dispatchEvent(e interface{}) {
+	switch e := e.(type) {
+	case *StatusMessage:
+		s.EmitStatusMessage(e)
+	case *AuctionMessage:
+		s.EmitAuctionMessage(e)
+	case *RfqMessage:
+		s.EmitRfqMessage(e)
+	case *TickerMessage:
+		s.EmitTickerMessage(e)
+	case *ReceivedLimitOrderMessage:
+		s.EmitReceivedLimitOrderMessage(e)
+	case *ReceivedMarketOrderMessage:
+		s.EmitReceivedMarketOrderMessage(e)
+	case *OpenMessage:
+		s.EmitOpenMessage(e)
+	case *DoneMessage:
+		s.EmitDoneMessage(e)
+	case *MatchMessage:
+		s.EmitMatchMessage(e)
+	case *AuthMakerMatchMessage:
+		s.EmitAuthMakerMatchMessage(e)
+	case *AuthTakerMatchMessage:
+		s.EmitAuthTakerMatchMessage(e)
+	case *StpChangeMessage:
+		s.EmitStpChangeMessage(e)
+	case *ModifyOrderChangeMessage:
+		s.EmitModifyOrderChangeMessage(e)
+	case *ActiveMessage:
+		s.EmitActiveMessage(e)
+	default:
+		log.Warnf("skip dispatching msg due to unknown message type: %T", e)
+	}
+}
+
+func createEndpoint(ctx context.Context) (string, error) {
+	return wsFeedURL, nil
+}
+
+type channelType struct {
+	Name       string   `json:"name"`
+	ProductIDs []string `json:"product_ids,omitempty"`
+}
+
+type websocketCommand struct {
+	Type       string        `json:"type"`
+	Channels   []channelType `json:"channels"`
+	Signature  *string       `json:"signature,omitempty"`
+	Key        *string       `json:"key,omitempty"`
+	Passphrase *string       `json:"passphrase,omitempty"`
+	Timestamp  *string       `json:"timestamp,omitempty"`
+}
+
+func (s *Stream) handleConnect() {
+	// subscribe to channels
+	if len(s.Subscriptions) == 0 {
+		return
+	}
+
+	subProductsMap := make(map[string][]string)
+	for _, sub := range s.Subscriptions {
+		strChannel := string(sub.Channel)
+		if _, ok := subProductsMap[strChannel]; !ok {
+			subProductsMap[strChannel] = []string{}
```

Comment:
`[]string{}` is redundant. We don't have to initialize slice since `append` handles everything.

![image](https://github.com/user-attachments/assets/b6ecf80a-9aa3-4aca-b8cd-8b457c9c881b)

Reply:
Done.